tck/src/main/java/org/eclipse/microprofile/context/tck/ManagedExecutorTest.java

```diff
@@ -1549,35 +1549,49 @@ public void newIncompleteFutureDependentStagesRunWithContext() throws ExecutionE
      */
     @Test
     public void propagateApplicationContext() throws ExecutionException, InterruptedException, TimeoutException {
+        ClassLoader appClassLoader = Thread.currentThread().getContextClassLoader();
+
         ManagedExecutor.Builder builder = ManagedExecutor.builder()
                 .propagated(ThreadContext.APPLICATION)
                 .cleared(ThreadContext.ALL_REMAINING);
 
-        ManagedExecutor executor;
-        try {
-            executor = builder.build();
-        }
-        catch (IllegalStateException x) {
-            return; // Skip test if Application context is not supported.
-        }
+        ManagedExecutor executor = builder.build();
 
+        ExecutorService unmanagedSingleThreadExecutor = Executors.newFixedThreadPool(1);
         try {
-            CompletableFuture<Class<?>> cf = executor.supplyAsync(() -> {
+            // Set the class loader of the single thread upon which unmanagedSingleThreadExecutor runs tasks
+            ClassLoader newClassLoader = unmanagedSingleThreadExecutor.submit(() -> {
+                ClassLoader loader = new ClassLoader() {};
+                Thread.currentThread().setContextClassLoader(loader);
+                return loader;
+            }).get(MAX_WAIT_NS, TimeUnit.NANOSECONDS);
+
+            // Use a ManagedExecutor to apply context to an operation that runs on unmanagedSingleThreadExecutor
+            CompletableFuture<Class<?>> cf = executor.completedFuture(1).thenApplyAsync(i -> {
                 try {
                     // load a class from the application
                     ClassLoader loader = Thread.currentThread().getContextClassLoader();
+                    Assert.assertEquals(loader, appClassLoader);
                     return loader.loadClass("org.eclipse.microprofile.context.tck.contexts.label.Label");
                 }
                 catch (ClassNotFoundException x) {
                     throw new CompletionException(x);
                 }
-            });
+            }, unmanagedSingleThreadExecutor);
 
             Assert.assertEquals(cf.get(MAX_WAIT_NS, TimeUnit.NANOSECONDS), Label.class,
                     "Could not load class from application's class loader.");
+
+            // Verify that the class loader of unmanagedSingleThreadExecutor was restored after
+            // running the previous task.
+            unmanagedSingleThreadExecutor.submit(() -> {
+                ClassLoader loader = Thread.currentThread().getContextClassLoader();
+                Assert.assertEquals(loader, newClassLoader);
```

Comment:
This assert needs to be moved to the "main" test thread. Exceptions that are thrown on other threads don't contribute to test status.

Reply:
I'm surprised to hear of a behavior like that so I tried it out by temporarily specify a value for the assert to intentionally make it fail.  The error reported in the test results for me just fine.  Nonetheless, I cannot rule out the possibility that it differs based on the environment in which you are running, so the safe thing to do is to move the asserts to the main test thread.  I have done that under commit https://github.com/eclipse/microprofile-context-propagation/pull/194/commits/f95439d40d62f35de3c9cd03fe6f1e2aa2778f71 - please review again.

Reply:
Weird, I've seen tests passing when a failing assert happened on an extra thread. But now I realize that was with JUnit, and this is TestNG, right? Perhaps TestNG takes care of that -- which would certainly be nice, I didn't know about that.

Reply:
Yeah, hopefully TestNG covers it better.  The other possibility I can think of would be if the failing assert on the separate thread happened _after_ the main test method completed.  That would never happen in our particular case because the main thread always waits for the completion stage that is running async to finish before exiting the test method.